src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php

```diff
@@ -607,16 +607,21 @@ function (Assertion $assertion) use ($generic_params) : Assertion {
                     return false;
                 }
             } elseif ($function->parts === ['define']) {
-                if ($first_arg && $first_arg->value instanceof PhpParser\Node\Scalar\String_) {
-                    $second_arg = $stmt->args[1];
-                    ExpressionAnalyzer::analyze($statements_analyzer, $second_arg->value, $context);
-                    $const_name = $first_arg->value->value;
-
-                    $statements_analyzer->setConstType(
-                        $const_name,
-                        isset($second_arg->value->inferredType) ? $second_arg->value->inferredType : Type::getMixed(),
-                        $context
-                    );
+                if ($first_arg) {
```

Comment:
NYC, not a new issue: One issue reported in Phan was that if there were multiple `define()` calls for a given constant (e.g. at src/a.php:6 and src/a.php:10), then the value the psalm infers would vary based on what file was being analyzed and the phase of analysis. I'm not sure if psalm would have a similar issue.

Another issue reported with Phan was dynamic constants, e.g. `$x = CONST_FROM_OTHER_FILE . 'suffix'; define($x, 'value');`. In that case, the constant name could only be resolved once all other files were parsed.

- those issues were resolved by allowing the first constant definition to be overridden only by a value from the same node (i.e. same file + line (+ column)).

Reply:
> One issue reported in Phan was that if there were multiple `define()` calls for a given constant (e.g. at src/a.php:6 and src/a.php:10), then the value the psalm infers would vary based on what file was being analyzed and the phase of analysis. I'm not sure if psalm would have a similar issue.

It does, including constants defined with `const`.